include/nbl/asset/utils/IShaderCompiler.h

```diff
@@ -223,108 +218,155 @@ class NBL_API2 IShaderCompiler : public core::IReferenceCounted
+				
+				// Lookup Hash is precompued at entry creation time. Even though in a preprocessing pass the compiler options' stage might change from a #pragma, 
+				// we can't update the info that the Cache uses to find entries post preprocessing: this would make it so that we need to preprocess entries
+				// to get the "real" stage before lookup in the cache, defeating its purpose
+				inline SEntry(const std::string_view _mainFileContents, const SCompilerOptions& compilerOptions)
+					: mainFileContents(std::move(std::string(_mainFileContents))), compilerArgs(compilerOptions)
 				{
+					// Form the hashable for the compiler data
+					size_t preprocessorArgsHashableSize = compilerArgs.preprocessorArgs.sourceIdentifier.size() + compilerArgs.preprocessorArgs.extraDefines.size() * sizeof(SMacroData);
+					auto stageString = std::to_string(compilerArgs.stage);
+					auto versionString = std::to_string(static_cast<uint32_t>(compilerArgs.targetSpirvVersion));
+					auto debugString = std::to_string(static_cast<uint8_t>(compilerArgs.debugInfoFlags.value));
+					size_t compilerArgsHashableSize = stageString.size() + versionString.size() + debugString.size() + compilerArgs.optimizerPasses.size();
+					std::vector<char> hashable;
+					hashable.reserve(preprocessorArgsHashableSize + compilerArgsHashableSize + mainFileContents.size());
+					
+					// Insert preproc stuff
+					hashable.insert(hashable.end(), compilerArgs.preprocessorArgs.sourceIdentifier.begin(), compilerArgs.preprocessorArgs.sourceIdentifier.end());
+					core::vector<SMacroData> sortedExtraDefines;
+					sortedExtraDefines.assign(compilerArgs.preprocessorArgs.extraDefines.begin(), compilerArgs.preprocessorArgs.extraDefines.end());
+					
+					// Sort them by identifier so the hash is not order-sensitive!
+					std::sort(sortedExtraDefines.begin(), sortedExtraDefines.end(), [](const SMacroData& lhs, const SMacroData& rhs) {return lhs.identifier < rhs.identifier; });
+					for (const auto& defines : sortedExtraDefines) {
+						hashable.insert(hashable.end(), defines.identifier.begin(), defines.identifier.end());
+						hashable.insert(hashable.end(), defines.definition.begin(), defines.definition.end());
+					}
+
+					// Insert rest of stuff from this struct
+					hashable.insert(hashable.end(), stageString.begin(), stageString.end());
+					hashable.insert(hashable.end(), versionString.begin(), versionString.end());
+					hashable.insert(hashable.end(), debugString.begin(), debugString.end());
+					for (auto pass : compilerArgs.optimizerPasses) {
+						hashable.push_back(static_cast<char>(pass));
+					}
+
+					// Now add the mainFileContents and produce both lookup and early equality rejection hashes
+					hashable.insert(hashable.end(), mainFileContents.begin(), mainFileContents.end());
+					std::string hashableString(std::move_iterator(hashable.begin()), std::move_iterator(hashable.end()));
+					lookupHash = std::hash<std::string>{}(hashableString);
```

Comment:
you don't need to convert everything to a string and use `std::hash` you were supposed to use the xxx256Hash because its quite a bit faster and can be applied over any byte array (doesn't care about `\0` string terminators)

Reply:
Also `std::hash` is size_t which is 64bit, but xx256Hash is 256 bit so less chance of collision

Reply:
[xxHash256](https://github.com/Devsh-Graphics-Programming/Nabla/blob/44c706dea466a546d6abaea3cede0dba4c52c924/include/nbl/core/xxHash256.h#L150)

Reply:
the 256bit hash could then be passed through std::hash to obtain a 64bit value for the functor that `unordered_map` needs

Reply:
[xxHash256](https://github.com/Devsh-Graphics-Programming/Nabla/blob/44c706dea466a546d6abaea3cede0dba4c52c924/include/nbl/core/xxHash256.h#L150)

Reply:
std::hash does not take an array, but boost::hash does. Do we have that available?